airflow/providers/amazon/aws/fs/s3.py

```diff
@@ -0,0 +1,131 @@
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+from __future__ import annotations
+
+import logging
+from functools import partial
+from typing import TYPE_CHECKING, Any, Callable, Dict
+
+import requests
+from botocore import UNSIGNED
+from requests import HTTPError
+
+from airflow.providers.amazon.aws.hooks.base_aws import AwsGenericHook
+
+if TYPE_CHECKING:
+    from botocore.awsrequest import AWSRequest
+    from fsspec import AbstractFileSystem
+
+
+Properties = Dict[str, str]
+
+S3_PROXY_URI = "s3.proxy-uri"
+
+log = logging.getLogger(__name__)
+
+schemes = ["s3", "s3a", "s3n"]
+
+
+class SignError(Exception):
+    """Raises when unable to sign a S3 request."""
+
+
+def get_fs(conn_id: str | None) -> AbstractFileSystem:
+    try:
+        from s3fs import S3FileSystem
+    except ImportError:
+        raise ImportError(
+            "Airflow FS S3 protocol requires the s3fs library, but it is not installed as it requires"
+            "aiobotocore. Please install the s3 protocol support library by running: "
+            "pip install apache-airflow[s3]"
+        )
+
+    if conn_id is None:
+        return S3FileSystem()
+
+    aws: AwsGenericHook = AwsGenericHook(aws_conn_id=conn_id)
+
+    client_kwargs = {
+        "endpoint_url": aws.conn_config.endpoint_url,
+        "aws_access_key_id": aws.conn_config.aws_access_key_id,
+        "aws_secret_access_key": aws.conn_config.aws_secret_access_key,
+        "aws_session_token": aws.conn_config.aws_session_token,
+        "region_name": aws.conn_config.region_name,
+    }
+    config_kwargs = {}
+    register_events: dict[str, Callable[[Properties], None]] = {}
+
+    if signer := aws.conn_config.extra_config.get("s3.signer"):
+        log.info("Loading signer %s", signer)
+        if singer_func := SIGNERS.get(signer):
+            uri = aws.conn_config.extra_config.get("s3.signer_uri")
+            token = aws.conn_config.extra_config.get("s3.signer_token")
+            if not uri or not token:
+                raise ValueError(f"Signer {signer} requires uri and token")
+
+            properties: Properties = {
+                "uri": uri,
+                "token": uri,
+            }
+            singer_func_with_properties = partial(singer_func, properties)
+            register_events["before-sign.s3"] = singer_func_with_properties
+
+            # Disable the AWS Signer
+            config_kwargs["signature_version"] = UNSIGNED
+        else:
+            raise ValueError(f"Signer not available: {signer}")
+
+    if proxy_uri := aws.conn_config.extra_config.get(S3_PROXY_URI):
+        config_kwargs["proxies"] = {"http": proxy_uri, "https": proxy_uri}
+
+    fs = S3FileSystem(client_kwargs=client_kwargs, config_kwargs=config_kwargs)
```

Comment:
`config_kwargs` looks the similar as `config_kwargs` from [AWS Connection Extras](https://airflow.apache.org/docs/apache-airflow-providers-amazon/stable/connections/aws.html#configuring-the-connection)

Reply:
is that added from `AwsGenericHook(config=config_kwargs)`?

Reply:
Not only.

Let's describe the order of obtaining [`botocore.config.Config`](https://botocore.amazonaws.com/v1/documentation/api/latest/reference/config.html)

1. Directly pass into the hook parameter
2. Otherwise try to build from `config_kwargs` new `botocore.config.Config`
3. Use default one

One important thing that this config uses during client creation not session, so even if we provide session directly we still need to pass `config_kwargs` parameters. 

And I guess that we can't grab this object and provide it into the S3FileSystem, but we could get this config from the connection by `aws.conn_config.extra_config.get("config_kwargs", {})` if it applicable into this place.

Reply:
Yes they are. I assume you mean we either need to update the documentation on the connection details or use the supplied `config_kwargs`. I think the first is better as doing config_kwargs by hand requires understanding the parameters and is a bit error prone.

Reply:
done